apis/v1alpha2/grpcroute_types.go

```diff
@@ -0,0 +1,461 @@
+	//
+	// * an unknown or unsupported kind of resource
+	// * a resource that does not exist
+	// * a resource in another namespace when the reference has not been
+	//   explicitly allowed by a ReferenceGrant (or equivalent concept).
+	//
+	// When a BackendRef is invalid, `UNIMPLEMENTED` statuses MUST be returned for
+	// requests that would have otherwise been routed to an invalid backend. If
+	// multiple backends are specified, and some are invalid, the proportion of
+	// requests that would otherwise have been routed to an invalid backend
+	// MUST receive an `UNIMPLEMENTED` status.
+	//
+	// When a BackendRef refers to a Service that has no ready endpoints, it is
+	// recommended to return an `UNAVAILABLE` status.
+	//
+	// Support: Core for Kubernetes Service
+	// Support: Custom for any other resource
+	//
+	// Support for weight: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	BackendRefs []GRPCBackendRef `json:"backendRefs,omitempty"`
+}
+
+// GRPCRouteMatch defines the predicate used to match requests to a given
+// action. Multiple match types are ANDed together, i.e. the match will
+// evaluate to true only if all conditions are satisfied.
+//
+// For example, the match below will match a gRPC request only if its service
+// is `foo` AND it contains the `version: v1` header:
+//
+// “`
+// matches:
+// - method:
+//
+//	  type: Exact
+//	  service: "foo"
+//	headers:
+//	- name: "version"
+//	  value "v1"
+//
+// “`
+type GRPCRouteMatch struct {
+	// Path specifies a gRPC request service/method matcher. If this field is not
+	// specified, all services and methods will match.
+	//
+	// +optional
+	// +kubebuilder:default={type: "Exact"}
+	Method *GRPCMethodMatch `json:"method,omitempty"`
+
+	// Headers specifies gRPC request header matchers. Multiple match values are
+	// ANDed together, meaning, a request must match all the specified headers
+	// to select the route.
+	//
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Headers []GRPCHeaderMatch `json:"headers,omitempty"`
+}
+
+// GRPCMethodMatch describes how to select a gRPC route by matching the gRPC
+// request service and/or method.
+//
+// At least one of Service and Method must be a non-empty string.
+type GRPCMethodMatch struct {
+	// Type specifies how to match against the service and/or method.
+	// Support: Core (Exact with service and method specified)
+	//
+	// Support Custom (Exact with method specified but no service specified)
+	//
+	// Support: Custom (RegularExpression)
+	//
+	// +optional
+	// +kubebuilder:default=Exact
+	Type *GRPCMethodMatchType `json:"type,omitempty"`
+
+	// Value of the service to match against. If left empty or omitted, will
+	// match all services.
+	//
+	// At least one of Service and Method must be a non-empty string.
+	// +optional
+	// +kubebuilder:default=""
+	// +kubebuilder:validation:MaxLength=1024
+	Service *string `json:"service,omitempty"`
```

Comment:
Can we add some regex validation here? What characters would be valid?

Reply:
Trying to get consensus from the gRPC community on what's reasonable here since this isn't codified. For the moment, I'm just disallowing the slash character.